datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -728,6 +731,73 @@ private void removeTaskNode(String instance, String name) {
+   * </ol>
+   */
+  public void updateAllAssignmentsAndIssueTokens(Map<String, List<DatastreamTask>> assignmentsByInstance,
+      List<DatastreamGroup> stoppingDatastreamGroups) {
+    Validate.notNull(assignmentsByInstance);
+    Validate.notNull(stoppingDatastreamGroups);
+
+    Map<String, Set<DatastreamTask>> currentAssignment = getAllAssignedDatastreamTasks();
+    Set<String> stoppingDatastreamTaskPrefixes = stoppingDatastreamGroups.stream().
+        map(DatastreamGroup::getTaskPrefix).collect(toSet());
+    Map<String, DatastreamGroup> taskPrefixDatastreamGroups = stoppingDatastreamGroups.stream().
+        collect(Collectors.toMap(DatastreamGroup::getTaskPrefix, Function.identity()));
+
+    // For each stopping datastream group, find all the instances that currently have tasks assigned
+    Map<DatastreamGroup, Set<String>> stoppingDgInstances = new HashMap<>();
+    for (String instance : currentAssignment.keySet()) {
+        Set<DatastreamTask> instanceTasks = currentAssignment.get(instance);
+        List<DatastreamTask> stoppingTasks = instanceTasks.stream().
+            filter(t -> stoppingDatastreamTaskPrefixes.contains(t.getTaskPrefix())).
+            collect(Collectors.toList());
+        for (DatastreamTask stoppingTask : stoppingTasks) {
+          DatastreamGroup datastreamGroup = taskPrefixDatastreamGroups.get(stoppingTask.getTaskPrefix());
+          stoppingDgInstances.computeIfAbsent(datastreamGroup, k -> new HashSet<>()).add(instance);
+        }
+      }
+
+    String hostname = "localhost";
+    try {
+      hostname = InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException ex) {
+      LOG.warn("Unable to obtain hostname for leader");
+    }
+
+    // Issue assignment tokens for each instance and each stopping datastream
+    for (DatastreamGroup stoppingGroup : stoppingDatastreamGroups) {
+      for (Datastream stoppingStream : stoppingGroup.getDatastreams()) {
+        String path = KeyBuilder.datastream(_cluster, stoppingStream.getName());
+
+        if (!_zkclient.exists(path)) {
+          LOG.warn("Trying to issue assignment tokens for non-existing stream: " + stoppingStream.getName());
+          continue;
+        }
+
+        String assignmentTokensPath = KeyBuilder.datastreamAssignmentTokens(_cluster, stoppingStream.getName());
+        _zkclient.ensurePath(assignmentTokensPath);
+
+        Set<String> instances = stoppingDgInstances.get(stoppingGroup);
+        LOG.info("Issuing assignment tokens for stream {} and {} instance(s)", stoppingStream.getName(), instances.size());
+        for (String instance : instances) {
+          String assignmentTokenPath = KeyBuilder.datastreamAssignmentTokenForInstance(_cluster,
+              stoppingStream.getName(), instance);
+          AssignmentToken token = new AssignmentToken(hostname, instance);
+          _zkclient.create(assignmentTokenPath, token.toJson(), CreateMode.PERSISTENT);
```

Comment:
could something bad happen here if two concurrent stop requests occur? Like should we have a check routine if an assignment token already exists for a host, then we don't create it?

Reply:
looks like `_zkclient.create` can throw exception, in that case are we okay with the partial assignmentToken written? Does leader know that there needs to be N assignmentToken assigned and there is only X, so even though those X have said that I have finished my assignments, leaders will not consider this as a success?

Reply:
@shrinandthakkar assignment token path is created under the datasream. Your previous work with datastream states already guarantees that no two concurrent stop requests on a datastream can happen at any given time in the cluster (state transitions are rejected if datastream is in STOPPING state).

@hshukla Zookeeper writes are atomic - there's no way we can end up with a partially written token. If the concern is more in the lines of "what if Zookeeper fails and leader writes subset of the tokens", then it's a cluster problem. One such example is session expiry. It is a critical failure and a lot of work went into handling it. No need for extra retry loops or checks here imo.

Reply:
> "what if Zookeeper fails and leader writes subset of the tokens",
Yes, I was referring this.  There could be a transient failures too does not mean the ZK is down. If there are 5 tokens to write, and it fails after 3, what happens here? will the request fail or if not, then  when leader polls the result does it consider that as a successful restart?

Reply:
@jzakaryan, as far as I understand, even after my work on ensuring that the leader would create an assignment when there's a stop request, there could still be a chance that two concurrent stop or restart requests occur. Still, since the reassignment is the same, the other request would be considered a no-op. 

So, my question here is if two concurrent stop requests occur, is there something that might create issues for the code in `updateAllAssignmentsAndIssueTokens` before it calls `updateAllAssignments`? 
1. What if a host claimed a token and a concurrent stop request created a new token for that same host until that host finished the request?  
2. ZK-related issues could also occur, but those would be cluster-level issues, which could not be handled here, I understand.

Reply:
@shrinandthakkar just to make sure I understand your concern, are you claiming that 2 concurrent assignment requests for the **same datastream** can occur?

Reply:
@hshukla Zookeeper client will throw a runtime exception and it will be caught by the `Coordinator` which will schedule a retry.
https://github.com/linkedin/brooklin/blob/71a9dea076531981cde7be52693cd98be1e34318/datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java#L1242

The only potential problem here is that the consecutive attempts at assignment change can result in some hosts ending up with more than one token. I'll make sure to handle that in Part 4 PR (cleanup and edge cases).

Reply:
That is exactly my concern: we need to have a no-op behavior implemented in the hosts handling these tokens if they are already working on stopping the producer for datastreams. If you are handling that in the upcoming PR, I am good with this.

Reply:
@shrinandthakkar let's sync up offline on this to make sure I understand everything, and I'll handle this in the edge cases PR